kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt

```diff
@@ -0,0 +1,217 @@
+          dialogs = emptyList()
+        }
+      }
+    }
+  }
+
+  override fun onBackPressed(): Boolean {
+    // This should only be hit if there are no dialogs showing, so we only
+    // need to consider the body.
+    return (body as? HandlesBack)?.onBackPressed() == true
+  }
+
+  fun takeScreens(
+    screens: Observable<out MainAndModalScreen<*, *>>,
+    viewRegistry: ViewRegistry
+  ) {
+    takeScreensSubs.clear()
+
+    // This looks like we're leaking subscriptions to screens, since we only unsubscribe
+    // if takeScreens() is called again. It's fine, though:  the switchMap in whileAttached()
+    // ensures that we're only subscribed to it while this view is attached to a window.
+
+    // Create a new body view each time the type of [MainAndModalScreen.main] changes.
+    takeScreensSubs.add(screens
+        .whileAttached()
+        .distinctUntilChanged { screen -> screen.main::class }
+        .subscribe {
+          removeAllViews()
+          addView(it.viewForMain(screens, viewRegistry, this))
+        }
+    )
+
+    // Compare the new modals list to the set of dialogs already on display. Show
+    // any new ones, throw away any stale ones.
+    takeScreensSubs.add(screens
+        .whileAttached()
+        .subscribe { mainAndModal ->
+          val newDialogs = mutableListOf<DialogRef>()
+          for ((i, modalScreen) in mainAndModal.modals.withIndex()) {
+            newDialogs +=
+              if (dialogs.size < i && dialogs[i].screen::class == modalScreen::class) {
+                dialogs[i]
+              } else {
+                DialogRef(modalScreen, mainAndModal.showDialog(i, screens, viewRegistry, this))
+              }
+          }
+
+          dialogs.filter { it !in newDialogs }
+              .forEach { it.dialog.hide() }
+          dialogs = newDialogs
+        }
+    )
+  }
+
+  private fun showAlert(alertScreen: AlertScreen): AlertDialog {
+    val builder = AlertDialog.Builder(context)
+
+    if (alertScreen.cancelable) {
+      builder.setOnCancelListener { alertScreen.onEvent(Canceled) }
+    } else {
+      builder.setCancelable(false)
+    }
+
+    for ((button, name) in alertScreen.buttons) {
+      when (button) {
+        POSITIVE -> builder.setPositiveButton(name) { _, _ ->
+          alertScreen.onEvent(ButtonClicked(POSITIVE))
+        }
+        NEGATIVE -> builder.setNegativeButton(name) { _, _ ->
+          alertScreen.onEvent(ButtonClicked(NEGATIVE))
+        }
+        NEUTRAL -> builder.setNeutralButton(name) { _, _ ->
+          alertScreen.onEvent(ButtonClicked(NEUTRAL))
+        }
+      }
+    }
+
+    alertScreen.message.takeIf { it.isNotEmpty() }
+        .let { builder.setMessage(it) }
+    alertScreen.title.takeIf { it.isNotEmpty() }
+        .let { builder.setTitle(it) }
+
+    return builder.show()
+  }
+
+  private fun <T> Observable<T>.whileAttached(): Observable<T> =
+    attached.switchMap { isAttached -> if (isAttached) this else never<T>() }
```

Comment:
Cute

Reply:
`switchMap()` seems like the solution to all the problems.